rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -204,4 +252,110 @@ void Player::prepare_publishers(const PlayOptions & options)
   }
 }
 
+void Player::handle_keypress()
+{
+  // Spin keyboard thread until all messages have been played
+  while (rclcpp::ok() && !played_all_) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(30));
+
+    int input = read_char_from_stdin();
+    switch (input) {
+      case ' ':
+        if (paused_) {
+          rosbag2_transport_->activate();
+        } else {
+          rosbag2_transport_->deactivate();
+        }
+        paused_ = !paused_;
+        break;
+    }
+  }
+}
+
+void Player::setup_terminal()
```

Comment:
Could that remaining part here be extracted in a separate `KeyboardHandler` or some other entity? I don't really see them necessarily being part of the `Player`.

Reply:
Agreed! I've seen a request for Action/Service control of rosbag playback, so if we abstract the controls into a common API that should be a lot easier to do as well.

Reply:
fd97872
Currently it is refactored to a separate `KeyboardHandler` class.
The Action/Service control would probably use interfaces associated with the LifecycleNode that Rosbag2Transport is now, so it will still want access to the `rosbag2_transport_` object in the player.
This may or may not be related. For some reason, I never see the Rosbag2Transport node listed in `ros2 node list` even before, nor now with `ros2 lifecycle nodes`. I don't know if that would affect the services. I was looking for the node to see if I can trigger a service in the tests, but not finding the node listed, I triggered it internally through new methods in Rosbag2Transport instead.